src/Kuzzle.ts

```diff
@@ -815,4 +846,26 @@ Discarded request: ${JSON.stringify(request)}`));
 
     dequeuingProcess();
   }
+
+  /**
+   * Sends a request with a timeout
+   */
+  private _timeoutRequest(delay, {request, options = {}}) {
+    // No timeout
+    if (delay === -1) {
+      return this.protocol.query(request, options);
+    }
+
+    const timeout = new Promise((resolve, reject) => {
+      const id = setTimeout(() => {
+        clearTimeout(id);
+        reject(`Request timed out after ${delay} ms`);
```

Comment:
Rejection should be done with a proper error class. 
IMHO a new error class must be created